lib/internal/streams/readable.js

```diff
@@ -1057,6 +1058,30 @@ Readable.prototype.wrap = function(stream) {
   return this;
 };
 
+Readable.prototype.collect = async function({
+  limit = Infinity,
+  rejectAtLimit = true
+} = {}) {
+  const bufs = [];
+  let len = 0;
+  const isObjectMode = this._readableState.objectMode;
+  const encoding = isObjectMode ? false : this._readableState.encoding;
+  for await (const chunk of this.iterator({ destroyOnReturn: rejectAtLimit })) {
+    const chunkLen = isObjectMode ? 1 : chunk.length;
+    if (len + chunkLen > limit) {
+      if (rejectAtLimit) {
+        throw new ERR_STREAM_COLLECT_LIMIT_EXCEEDED(len + chunk.length, limit);
+      } else {
+        this.unshift(chunk);
+        break;
+      }
+    }
+    len += chunkLen;
+    bufs.push(chunk);
+  }
+  return isObjectMode ? bufs : encoding ? bufs.join('') : Buffer.concat(bufs);
```

Comment:
Hmm.

What if `limit` is actually `Math.min(limit, buffer.kMaxLength)` and it's documented (and tested) as such?

Reply:
About the limit, I'd prefer we throw a range error instead of silently capping at `buffer.kMaxLength` (we should also throw if limit is negative or not an integer)

An exception is still possible if there isn't enough memory to allocate the buffer.

Reply:
That would work also, I think.